OpenRA.Mods.D2k/Traits/Sandworm.cs

```diff
@@ -0,0 +1,130 @@
+#region Copyright & License Information
+/*
+* Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+* This file is part of OpenRA, which is free software. It is made
+* available to you under the terms of the GNU General Public License
+* as published by the Free Software Foundation. For more information,
+* see COPYING.
+*/
+#endregion
+
+using System;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.D2k.Traits
+{
+	class SandwormInfo : WandersInfo, Requires<MobileInfo>, Requires<RenderUnitInfo>, Requires<AttackBaseInfo>
+	{
+		[Desc("Time between rescanning for targets (in ticks).")]
+		public readonly int TargetRescanIntervalInTicks = 32;
+
+		[Desc("The radius in which the worm \"searches\" for targets.")]
+		public readonly int MaxSearchRadius = 32;
+
+		[Desc("The range at which the worm launches an attack regardless of noise levels.")]
+		public readonly int LaunchAttackRange = 3;
+
+		public readonly string IdleSequenceName = "idle";
+
+		public override object Create(ActorInitializer init) { return new Sandworm(init.Self, this); }
+	}
+
+	class Sandworm : Wanders, ITick, INotifyBecomingIdle
+	{
+		readonly Lazy<Mobile> mobile;
+		readonly Lazy<RenderUnit> renderUnit;
+		readonly Lazy<AttackBase> attackTrait;
+		readonly SandwormInfo info;
+
+		public bool IsAttacking;
+
+		int targetCountdown;
+
+		public Sandworm(Actor self, SandwormInfo info)
+			: base(self, info)
+		{
+			this.info = info;
+			mobile = Exts.Lazy(self.Trait<Mobile>);
+			renderUnit = Exts.Lazy(self.Trait<RenderUnit>);
+			attackTrait = Exts.Lazy(self.Trait<AttackBase>);
+		}
+
+		public new void OnBecomingIdle(Actor self)
+		{
+			if (renderUnit.Value.DefaultAnimation.CurrentSequence.Name != info.IdleSequenceName)
+				renderUnit.Value.DefaultAnimation.ReplaceAnim("idle");
+
+			base.OnBecomingIdle(self);
+		}
+
+		public override void DoAction(Actor self, CPos targetCell)
+		{
+			if (AnyValidTargets(self))
+				return;
+
+			Game.Debug("(Sandworm {0}) is wandering.", self.ActorID);
+			self.QueueActivity(mobile.Value.MoveWithinRange(Target.FromCell(self.World, targetCell, SubCell.Any), WRange.FromCells(1)));
+		}
+
+		public void Tick(Actor self)
+		{
+			if (--targetCountdown > 0 || IsAttacking || !self.IsInWorld)
+				return;
+
+			AnyValidTargets(self);
+		}
+
+		bool AnyValidTargets(Actor self)
```

Comment:
The logic here really needs to be split up.
Either put it all into tick or make `AnyValidTargets` return a valid state that the rest of the logic depends on.